### PR TITLE
[GTK4 Prep] Move External App Actions Logic To Utils

### PR DESCRIPTION
**Pull request by colinkiama**
Part of of "Replace Gtk.MenuItem with GLib.MenuModel" task in #1157

### CODE REVIEW

File: src/Utils.vala [hunk 1]
```diff
@@ -184,4 +184,109 @@ namespace Scratch.Utils {
...
+        var file = GLib.File.new_for_path (path);
+        var file_list = new List<GLib.File> ();
+        file_list.append (file);
+        var app_info = new GLib.DesktopAppInfo (app_id);
+        try {
+            app_info.launch (file_list, null);
+        } catch (Error e) {
+            warning (e.message);
+        }
+    }
```
Comment (danirabbit) [inline src/Utils.vala:246]: `file_type` appears to be unused here. This seems to be a duplicate of `launch_app_with_file` but just with the one liner to create a GLib.File from path. Should this function exist?

File: src/Utils.vala [hunk 2]
```diff
@@ -184,4 +184,109 @@ namespace Scratch.Utils {
...
+                    GLib.VariantType.STRING,
+                    { file.get_path (), app_id, file_type }
+                )
+            };
+
+            menuitem.add (menuitem_grid);
+            menu.add (menuitem);
+        }
+    }
+
```
Comment (danirabbit) [inline src/Utils.vala:223]: extra space

File: src/Utils.vala [hunk 3]
```diff
@@ -184,4 +184,109 @@ namespace Scratch.Utils {
...
+                menu.add (menu_item);
+            }
+        } catch (Error e) {
+            warning (e.message);
+        }
+
+        return menu;
+    }
+
```
Comment (danirabbit) [inline src/Utils.vala:272]: extra space

**Review by danirabbit — CHANGES_REQUESTED** [review]
Seems sensible. Just one question and a couple comments about extra whitespace

File: src/Utils.vala (hunk 1 above)
Comment (colinkiama) [reply src/Utils.vala]: No, it doesn't need to. It has been removed now

**Comment by colinkiama** [comment]
I've made new changes that should resolve all the issues now.

File: src/Utils.vala [hunk 4]
```diff
@@ -184,4 +184,94 @@ namespace Scratch.Utils {
...
+        } catch (Error e) {
+            warning (e.message);
+        }
+    }
+
+    public Gtk.Menu create_contract_items_for_file (GLib.File file, string file_type) {
+        var menu = new Gtk.Menu ();
+
+        try {
+            var contracts = Granite.Services.ContractorProxy.get_contracts_by_mime (file_type);
```
Comment (danirabbit) [inline src/Utils.vala:239]: Should this be https://valadoc.org/granite-7/Granite.Services.ContractorProxy.get_contracts_for_file.html so we don't have to manually get file type?

File: src/Utils.vala [hunk 5]
```diff
@@ -184,4 +184,94 @@ namespace Scratch.Utils {
...
+        }
+
+        return menu;
+    }
+
+    public void execute_contract_with_file_path (string path, string contract_name, string file_type) {
+        var file = GLib.File.new_for_path (path);
+
+        try {
+            var contracts = Granite.Services.ContractorProxy.get_contracts_by_mime (file_type);
```
Comment (danirabbit) [inline src/Utils.vala:264]: Same here

File: src/Utils.vala (hunk 4 above)
Comment (colinkiama) [reply src/Utils.vala]: Yeah. That makes the code more concise. Thanks!

File: src/Utils.vala (hunk 5 above)
Comment (colinkiama) [reply src/Utils.vala]: This has been updated now too.

**Review by danirabbit — APPROVED** [review]
Nice work!

**Pull request merged**